tools/repo-utils/backport.sh

```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+
+# Check if the correct number of arguments are provided
+if [ "$#" -ne 2 ]; then
+  echo "Usage: $0 <commit-sha> <target-branch>"
+  exit 1
+fi
+
+# assign input arguments to variables
+COMMIT_SHA=$1
+TARGET_BRANCH=$2
+NEW_BRANCH="backport-${COMMIT_SHA}"
+REPOSITORY="https://api.github.com/repos/fleetdm/fleet"
+
+# get_github_token tries to get a token using the gh CLI
+get_github_token_from_cli() {
+  if ! command -v gh &> /dev/null
+  then
+    echo "gh CLI could not be found, please install it or provide a GitHub token."
+    exit 1
+  fi
+
+  GH_TOKEN=$(gh auth token)
+  if [ -z "$GH_TOKEN" ]; then
+    echo "Failed to retrieve GitHub token using gh CLI."
+    exit 1
+  fi
+  echo $GH_TOKEN
+}
+
+# check if the GITHUB_TOKEN environment variable is set
+if [ -z "$GITHUB_TOKEN" ]; then
+  echo "GitHub token not provided, attempting to retrieve it using gh CLI..."
+  GITHUB_TOKEN=$(get_github_token_from_cli)
+fi
+
+# check if the GitHub token is still empty
+if [ -z "$GITHUB_TOKEN" ]; then
+  echo "Error: GitHub token is still empty. Exiting."
+  exit 1
+fi
+
+# grab the SHA of the RC branch
+BASE_SHA=$(curl -s -H "Authorization: token $GITHUB_TOKEN" -H "Accept: application/vnd.github.v3+json" $REPOSITORY/git/refs/heads/$TARGET_BRANCH | jq -r .object.sha)
+
+# create a new branch using the RC as the base branch
+curl -s -X POST -H "Authorization: token $GITHUB_TOKEN" -H "Accept: application/vnd.github.v3+json" $REPOSITORY/git/refs \
+  -d "{\"ref\": \"refs/heads/${NEW_BRANCH}\", \"sha\": \"${BASE_SHA}\"}" > /dev/null 2>&1
+
+# create a new commit with the contents of the original commit
+COMMIT_INFO=$(curl -s -H "Authorization: token $GITHUB_TOKEN" -H "Accept: application/vnd.github.v3+json" $REPOSITORY/commits/${COMMIT_SHA})
+TREE_SHA=$(echo $COMMIT_INFO | jq -r .commit.tree.sha)
+COMMIT_TITLE=$(echo $COMMIT_INFO | jq -r .commit.message)
+NEW_COMMIT_SHA=$(curl -s -X POST -H "Authorization: token $GITHUB_TOKEN" -H "Accept: application/vnd.github.v3+json" $REPOSITORY/git/commits \
+  -d "{\"message\": \"Backport: ${COMMIT_TITLE}\", \"tree\": \"${TREE_SHA}\", \"parents\": [\"${BASE_SHA}\"]}" | jq -r .sha)
+
+# associate that commit with the branch
+curl -s -X PATCH -H "Authorization: token $GITHUB_TOKEN" -H "Accept: application/vnd.github.v3+json" $REPOSITORY/git/refs/heads/${NEW_BRANCH} \
+  -d "{\"sha\": \"${NEW_COMMIT_SHA}\", \"force\": true}" > /dev/null 2>&1
+
+#  create a PR
+PR_TITLE="backport: ${COMMIT_TITLE}"
+PR_BODY="This is a backport of commit ${COMMIT_SHA} to ${TARGET_BRANCH}."
+
+PR_URL=$(curl -s -X POST -H "Authorization: token $GITHUB_TOKEN" -H "Accept: application/vnd.github.v3+json" $REPOSITORY/pulls \
+  -d "{\"title\":\"${PR_TITLE}\", \"body\":\"${PR_BODY}\", \"head\":\"${NEW_BRANCH}\", \"base\":\"${TARGET_BRANCH}\"}" | jq .html_url)
+
+echo "Successfully created backport branch and pull request: $PR_URL"
```

Comment:
Hmm I gave this a try and it exited with this:

```
$ ./backport.sh ef07a406cc8dbccb4914bbd27631df74e5b13b49 minor-fleet-v4.55.0
GitHub token not provided, attempting to retrieve it using gh CLI...
Successfully created backport branch and pull request: null
```

It did create the branch, but not the PR, and when I go ahead and create the PR it seems like both are identical. Not sure what step went wrong exactly though.

Reply:
Likely something between lines 51 and 60.

Reply:
thank you for giving it a try!! 🤗  I have been testing it with a private repo, will see if something is different when using `fleetdm/fleet`

Reply:
it fails because the message has double quotes! we should account for this case and also fail the script if one of the requests fails. TY!!

```
 -d '{"message": "Backport: Bugfix: list host software "Available for install" filter must show installers that have an install request on the host (#21083)", "tree": "5128a2a70d34654de0b875ce0d77ce87e874d23b", "parents": ["64e97b3a66f954e8ba92242a66990890375dfec3"]}'
```

Reply:
Ahhh! Makes sense! Yeah I think the `set -euo pipefail` would've failed the script at that point (assuming curl did return an error exit code). There may be an easy way to JSON-encode the message with `jq` so that special chars don't break the commands?

Reply:
Looks like this would do it?

```
$ echo 'An "Available" filter' | jq -R '.'
"An \"Available\" filter"
```